### ISSUE TITLE
[feature request] go-memdb backend

### ISSUE THREAD

**Issue by tao12345666333** [k3s-io/kine#97]
I want to know if we can add a new backend using `go-memdb` as in-memory database.

https://github.com/hashicorp/go-memdb

> Provides the `memdb` package that implements a simple in-memory database
> built on immutable radix trees. The database provides Atomicity, Consistency
> and Isolation from ACID. Being that it is in-memory, it does not provide durability.
> The database is instantiated with a schema that specifies the tables and indices
> that exist and allows transactions to be executed.
> 
> The database provides the following:
> 
> * Multi-Version Concurrency Control (MVCC) - By leveraging immutable radix trees
>   the database is able to support any number of concurrent readers without locking,
>   and allows a writer to make progress.
> 
> * Transaction Support - The database allows for rich transactions, in which multiple
>   objects are inserted, updated or deleted. The transactions can span multiple tables,
>   and are applied atomically. The database provides atomicity and isolation in ACID
>   terminology, such that until commit the updates are not visible.
> 
> * Rich Indexing - Tables can support any number of indexes, which can be simple like
>   a single field index, or more advanced compound field indexes. Certain types like
>   UUID can be efficiently compressed from strings into byte indexes for reduced
>   storage requirements.
> 
> * Watches - Callers can populate a watch set as part of a query, which can be used to
>   detect when a modification has been made to the database which affects the query
>   results. This lets callers easily watch for changes in the database in a very general
>   way.

**Comment by brandond** [comment]
That might be hard, since kine expects a SQL compatible storage engine. Seems like if you want an ephemeral dataatore you could get the same result out of sqlite with the special :memory: database path?

**Comment by tao12345666333** [comment]
yep, at the same time I also hope not to introduce other components to accomplish this.

**Comment by tao12345666333** [comment]
maybe we can using https://github.com/araddon/qlbridge to translate it. https://github.com/araddon/qlbridge/tree/master/datasource/memdb

**Comment by tokers** [comment]
Introduce an SQLite with the in-memory file path is still a bit heavy for us, we need some further discussions.

**Comment by brandond** [comment]
What specifically are you trying to achieve?

**Comment by brandond** [comment]
fwiw, I tried using sqlite in memory mode and it's pretty flakey, due to changes in table locking semantics.
```
docker run --rm -it --privileged --name k3s-server-1 rancher/k3s server --datastore-endpoint 'sqlite://file:memdb1?mode=memory&cache=shared'
```

**Comment by akutz** [comment]
> You could just use kine but store the SQLite database in a temporary directory that's not persistent?

True, but part of the goal is a super portable, self-contained means of implementing controllers in CLIs. I'll take your suggestion under advisement. Based on the response, can I assume that in-memory sqlite still does not work well?

**Comment by brandond** [comment]
You're welcome to try it and see if it works for you. It didn't work very well for a whole Kubernetes cluster using the example command I gave.

**Comment by akutz** [comment]
Ack. Thank you. So just to be clear, to your knowledge no one has a) tried it since you did last July or b) worked on it since then either?

**Comment by brandond** [comment]
yep

**Comment by brandond** [comment]
Seems to be the same issue described here: https://github.com/mattn/go-sqlite3/issues/50#issuecomment-87811607

**Comment by mhmxs** [comment]
> That might be hard, since kine expects a SQL compatible storage engine.

@brandond Is this statement still valid? 

The reason I think it may not, is: here https://github.com/k3s-io/kine/blob/master/pkg/endpoint/endpoint.go#L59 `kine` creates a raw etcd listener without any sql specific wrapper. It looks simple thing to create any backend, not just SQL. As i see a backend only has to implement this interface: https://github.com/k3s-io/kine/blob/master/pkg/server/types.go#L20 and has to start it's own listener.

:pray:

**Comment by brandond** [comment]
We now have another non-sql backend in https://github.com/k3s-io/kine/tree/master/pkg/drivers/nats, but it was fairly non-trivial for the Synadia folks to implemement. You might look at that as an example of a non-sql store though.
